lib/ferrum/browser/subscriber.rb

```diff
@@ -17,7 +17,7 @@ def initialize
       end
 
       def on(event, &block)
-        @on[event] << block
+        @on[event] = [block]
```

Comment:
@route we have 2 subscribers for `Network.requestWillBeSent`, do we have missed test for this?

Reply:
Yes we should have an ability to define a few handlers.

* `Network.requestWillBeSent` in `frames.rb` is defined because we want to know if Chrome is going to make request for the main frame for the whole document (the whole page), if it does then we reset `@event` because now we shouldn't return back to the main Thread until page is loaded which signaled by Chrome.
* `Network.requestWillBeSent` in `network.rb` is used to define request and exchange instances that are used by people listening to network traffic

So in other words if you leave only one you won't be able to build correctly either network traffic or you will return immediately even though page is only started to load or reload.

Reply:
I don't see that it's wrong in any sense to have an ability to have a few handlers. I think that current implementation of authentication needs reconsidering, though it will be hard because we only use what Chrome provides.

Reply:
A lot of stuff is missing in current version of PR, it's not ready in any sense. I'm closing so that people don't use it because it breaks the gem, while tests are passing.

Reply:
> we should have an ability to define a few handlers

well, okay, then we may, probably, use the `redefine:` flag for `on` method to allow both cases:
- define a few handlers
- redefine existed handler

Reply:
so, with the flag usage, we can keep the current behavior of handlers is unchanged, and provide the fix for #149 in this way:
```
browser.on(:request, redefine: true) do |request|
  if allowed_types.include?(request.resource_type)
    puts "ALLOW #{request.resource_type} #{request.url}"
    request.continue
  else
    puts "DENY #{request.resource_type} #{request.url}"
    request.abort
  end
end
```
@route WDYT?

Reply:
Yep let me sleep with it for tomorrow.

Reply:
Decided to go with https://github.com/rubycdp/ferrum/pull/154, still not finished but I think that would work